lib/posix-fd: Move open file descriptions into own lib

andreittr opened the pull request:
### Description of changes

This PR moves the POSIXey concept of "open file descriptions" out of `lib/ukfile` and into its own library, merging the "common" definitions from `<uk/ofile.h>` with the POSIXey definitions from `<uk/posix-fd.h>`.
This allows decoupling `posix-fdio` from `posix-fdtab` and simplifies internal usage of anonymous (i.e., not managed by any fdtab) open file descriptions.

### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): [e.g. `x86_64` or N/A]
 - Platform(s): [e.g. `kvm`, `xen` or N/A]
 - Application(s): [e.g. `app-python3` or N/A]


### Additional configuration

N/A

andreittr committed:
lib/posix-fd: Move open file headers into own lib

This changes merges the headers uk/ofile.h and uk/posix-fd.h and moves
them into their own library, providing shared definitions related to
open file descriptions for both posix-fdio and posix-fdtab, without
introducing an undue dedependency between the two.

Checkpatch-Ignore: REPEATED_WORD
Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr committed:
lib/posix-fdio: Remove posix-fdtab dependency

This change removes the hard dependency on posix-fdtab, allowing
posix-fdio to be used with anonymous open file descriptions (i.e., raw
uk_ofiles). It remains selected as a soft dependency by default.
File descriptor related syscalls are now only provided if posix-fdtab is
also selected.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

mogasergiu commented on lib/posix-fd/include/uk/posix-fd.h:
```diff
@@ -12,7 +12,31 @@
+};
+
+static inline
+void uk_ofile_init(struct uk_ofile *of,
```
I see this as public, maybe should be commented with params/return as well?

mogasergiu commented on lib/posix-fd/Makefile.uk:
```diff
@@ -0,0 +1,4 @@
+$(eval $(call addlib_s,libposix_fd,$(CONFIG_LIBPOSIX_FD)))
+
+CINCLUDES-$(CONFIG_LIBPOSIX_FDIO) += -I$(LIBPOSIX_FD_BASE)/include
```
Did you mean`CONFIG_LIBPOSIX_FD`?

andreittr replied:
Ooof, that's an embarassing typo :sweat_smile: thanks for catching!

mogasergiu commented on lib/posix-fdio/Makefile.uk:
```diff
@@ -6,21 +6,21 @@ CXXINCLUDES-$(CONFIG_LIBPOSIX_FDIO) += -I$(LIBPOSIX_FDIO_BASE)/include
-LIBPOSIX_FDIO_SRCS-y += $(LIBPOSIX_FDIO_BASE)/fd-shim.c
+LIBPOSIX_FDIO_SRCS-$(CONFIG_LIBPOSIX_FDTAB) += $(LIBPOSIX_FDIO_BASE)/fd-shim.c
 
-UK_PROVIDED_SYSCALLS-$(CONFIG_LIBPOSIX_FDIO) += preadv2-5
```
Note that [provided_syscalls.awk](https://github.com/unikraft/unikraft/blob/staging/lib/syscall_shim/syscall_provided.awk) may still autogenerate these syscalls in `build/libsyscall_shim/provided_syscalls.in`.

I recommend

```diff
diff --git a/lib/posix-fdio/Makefile.uk b/lib/posix-fdio/Makefile.uk
index fb046a26..d2e62789 100644
--- a/lib/posix-fdio/Makefile.uk
+++ b/lib/posix-fdio/Makefile.uk
@@ -8,6 +8,7 @@ LIBPOSIX_FDIO_SRCS-y += $(LIBPOSIX_FDIO_BASE)/fdstat.c
 LIBPOSIX_FDIO_SRCS-y += $(LIBPOSIX_FDIO_BASE)/fdctl.c
 LIBPOSIX_FDIO_SRCS-$(CONFIG_LIBPOSIX_FDTAB) += $(LIBPOSIX_FDIO_BASE)/fd-shim.c
 
+ifeq ($(CONFIG_LIBPOSIX_FDIO), y)
 UK_PROVIDED_SYSCALLS-$(CONFIG_LIBPOSIX_FDTAB) += preadv2-5
 UK_PROVIDED_SYSCALLS-$(CONFIG_LIBPOSIX_FDTAB) += preadv-4
 UK_PROVIDED_SYSCALLS-$(CONFIG_LIBPOSIX_FDTAB) += pread64-4
@@ -24,3 +25,4 @@ UK_PROVIDED_SYSCALLS-$(CONFIG_LIBPOSIX_FDTAB) += fstat-2
 
 UK_PROVIDED_SYSCALLS-$(CONFIG_LIBPOSIX_FDTAB) += fcntl-3
 UK_PROVIDED_SYSCALLS-$(CONFIG_LIBPOSIX_FDTAB) += ioctl-3
+endif
```
Or, maybe even better, use FDIO for the -y and FDTAB for the one guard.


For example, this would lead in a build error in the case of having syscall shim and fdtab enabled, but fdio disabled.

andreittr replied:
Good catch, thanks! Darn awk scripts :smile: 
I'll also check if other libs do similar things and push update.

andreittr committed:
lib/posix-*: Makefile: guard syscalls behind fdtab

This change adds configuration guards around syscall declarations in
Makefiles that depend on posix-fdtab to be enabled. This allows builds
of these libs with syscall_shim but fdtab disabled, decoupling them from
the fdtab completely.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr commented:
Rebased to current staging

mogasergiu reviewed (approved):
All fine by me. Thanks, merging.

Reviewed-by: Sergiu Moga <sergiu@unikraft.io>
Approved-by: Sergiu Moga <sergiu@unikraft.io>
